### ISSUE TITLE
Ordnerstruktur änderbar ohne ID zu ändern

### ISSUE THREAD

**Issue by ldittmar81** [ioBroker/ioBroker.js-controller#905]
Ein der Gründe ein Alias zu erzeugen, ist dass die ID immer gleich bleibt, auch wenn das Gerät sich ändert. Leider lässt diese Tatsache auch nicht zu, dass man die Datenstruktur unterhalb von Alias ändert, denn diese ist ja Teil der ID.

Es wäre cool, wenn man die Struktur ändern könnte, aber ohne die ID zu ändern. Dazu würde im DP noch ein Feld "folder" hinzugefügt, der dazu dient den aktuellen Ordner für die Anzeige zu ermitteln. Bsp.: Die ID ist "alias.0.Licht.Küche.SET" im folder könnte aber stehen Licht.Erdgeschoss.Küche das heisst, dass der DP mit der ID alias.0.Licht.Küche.SET im Ordner Licht.Erdgeschoss.Küche sichtbar wäre.

**Comment by Diginix** [comment]
Das ist keine schlechte Idee nur die ObjektID selbst persistent zu haben, nicht aber ihren Pfad der Ablage mit im Skript zu benötigen. Jedes Objekt erhält eine UID und nur diese wird im Skript verwendet. Das Objekt kann unterhalb von Alias oder 0_userdata.0 aber liegen wo es will.
So könnte man später umsortieren usw. ohne Skripte nochmal anfassen zu müssen.
Unter alias.0 müsste man ansonsten komplett auf eine Struktur verzichten und tatsächlich für jedes 0_userdata Objekt welches man ggf. mal verschiebt ein alias Objekt erstellen.

**Comment by Apollon77** [comment]
Und der "Name" des Objekts kann das nicht schon abbilden?

**Comment by ldittmar81** [comment]
Leider nicht. Es geht um die gesamte Ordnerstruktur. Wenn die mal steht, dann können Datenpunkte nicht mehr bewegt werden. Das ist aber nur für die Ordnungsfanatiker wichtig.
Beispiel: Ich habe normale Lichter und RGBS alle zusammen unter Haus.Lichter ... wenn es irgendwann doch zuviele werden, möchte ich vielleicht diese unterteilen in Haus.Lichter.Normal und Haus.Lichter.RGB .... dann fällt mir ein, dass RGB nicht die richtige Bezeichnung wäre, und ich hätte lieber Haus.Lichter.Deko_Beleuchtung .... das ist das Problem

**Comment by AlCalzone** [comment]
Ist das nicht noch verwirrender, wenn man aus der Struktur nicht mehr die ID des Objekts ablesen kann, um sie z.B. in Skripten zu verwenden?

**Comment by ldittmar81** [comment]
Jain... Beim Blockly wird automatisch den Namen des Objektes angezeigt. Da ist die ID also egal... Beim Skripten hat man diesen Komfort nicht. Hmm... Man könnte den Skript Adapter so anpassen, dass es beim Mausover den Namen der ID zeigt... Außerdem wäre es durch möglich als selector [adapter].[instanz].[folder].[name] einzugeben. Also: ID = alias.0.sfhjksfkhsjdhfkjhfskd als Selector für die Skript-Programmierung "alias.0.haus.party.Wand Beleuchtung" ....Hmm... Theoretisch könnte man die aktuelle Version so lassen und ein neues Feld UID hinzufügen. Dann wäre die ID änderbar, aber das Objekt hätte noch eine nicht änderbare ID, die IMMER gültig ist.

**Comment by paul53** [comment]
Beim Skripten hat man den Komfort sehr wohl, wenn man die ID per Select-ID einfügt: Als Kommentar hinter der ID. Wenn über den Namen zugegriffen werden soll, kann man es mittels _getIdByName(name)_ machen, was allerdings eindeutige Namen voraussetzt.

**Comment by ldittmar81** [comment]
> Beim Skripten hat man den Komfort sehr wohl, wenn man die ID per Select-ID einfügt: Als Kommentar hinter der ID. 
Der Kommentar wird aber nicht aktualisiert, wenn der Name des Objektes sich ändert...

Hmm... damit die Kompatibilität zum alten bleibt, könnte man doch wirklich ein neuen Feld UID hinzufügen und dafür die eigentliche ID bearbeitbar machen, oder? Würde was dagegen sprechen? Wenn es dann soweit wäre, würde die neue Version des JS-Controllers zu jedem Objekt eine UID hinzufügen bestehend aus adapter.instanz.UID ... also alias.0.njf8778r4n3fn3hqhq. Diese würde sich nie wieder ändern und man könnte dann die normale ID weiterhin so nutzen, nur dass man es dann auch ändern kann. Alle Skripte die auf diese UID aufgebaut werden, würden dann immer funktionieren.

**Comment by AlCalzone** [comment]
> Kompatibilität

Das wäre einer der tiefsten Eingriffe in die Objektstruktur von ioBroker schlechthin. Die derzeitige ID **ist** die UID. 

Sinnvoller finde ich da eher den Vorschlag aus deinem OP, ein weiteres Feld bei der Anzeige im Admin hinzuzuziehen. Verwirrend fände ich es aber trotzdem, wenn dieses Feld z.B. von Adaptern automatisch gesetzt wird, ohne dass man das als User sieht oder optisch eindeutig erkennen kann.

**Comment by ldittmar81** [comment]
Klar... ein Feld "folder", der nur für die Ansicht zuständig ist, ist die "einfache" Variante. Das Feld müsste dann nicht zwingend da sein und nur beim Verschieben von Objekten würde es rein kommen. Ist zwar auch viel Arbeit, aber nicht so kritisch...
